core/modules/util/EventThread.h

```diff
@@ -0,0 +1,195 @@
+ *
+ *  @author: John Gates,
+ */
+
+#ifndef LSST_QSERV_UTIL_EVENTTHREAD_H_
+#define LSST_QSERV_UTIL_EVENTTHREAD_H_
+
+// System headers
+#include <atomic>
+#include <cassert>
+#include <deque>
+#include <thread>
+#include <vector>
+
+// Qserv headers
+#include "util/Command.h"
+
+namespace lsst {
+namespace qserv {
+namespace util {
+
+/// A queue of Commands meant to drive an EventThread.
+///
+class CommandQueue {
+public:
+    using Ptr = std::shared_ptr<CommandQueue>;
+    virtual ~CommandQueue() {};
+    /// Queue a command object in a thread safe way and signal any threads
+    /// waiting on the queue that a command is available.
+    virtual void queCmd(Command::Ptr const& cmd) {
+        std::lock_guard<std::mutex> lock(_mx);
+        _qu.push_back(cmd);
+        notify(_qu.size() > 1);
+    };
+
+    /// Get a command off the queue.
+    /// If wait is true, wait until a message is available.
+    virtual Command::Ptr getCmd(bool wait=true) {
+        std::unique_lock<std::mutex> lock(_mx);
+        if (wait) {
+            _cv.wait(lock, [this](){return !_qu.empty();});
+        }
+        if (_qu.empty()) {
+            return nullptr;
+        }
+        auto cmd = _qu.front();
+        _qu.pop_front();
+        return cmd;
+    };
+
+    /// Notify all threads waiting on this queue, or just 1 if all is false.
+    virtual void notify(bool all=true) {
+        if (all) {
+            _cv.notify_all();
+        } else {
+            _cv.notify_one();
+        }
+    }
+
+    virtual void commandStart(Command::Ptr const&) {};
+    virtual void commandFinish(Command::Ptr const&) {};
+protected:
+    std::deque<Command::Ptr> _qu{};
+    std::condition_variable  _cv{};
+    std::mutex               _mx{};
+};
+
+/// An event driven thread.
+/// Thread must be started with run(). Stop the thread by calling queEnd().
+class EventThread {
+public:
+    enum { HALT = -1000 };
+    EventThread() {}
+    explicit EventThread(CommandQueue::Ptr const& q) : _q{q} {}
+    EventThread(EventThread const&) = delete;
+    EventThread operator=(EventThread const&) = delete;
+    virtual ~EventThread() {}
+
+    void handleCmds();
+    void join() { _t.join(); }
+    void run(); ///< Run this EventThread.
+
+    /// Put a Command on this EventThread's queue.
+    void queCmd(Command::Ptr cmd) {
+        assert(cmd != nullptr);
+        _q->queCmd(cmd);
+    }
+
+    /// Queues and action that will stop the EventThread that answers it.
+    virtual void queEnd() {
+        struct MsgEnd : public Command {
+            int action() override { return HALT; }
+        };
+        std::shared_ptr<MsgEnd> cmd = std::make_shared<MsgEnd>();
+        queCmd(cmd);
+    }
+
+protected:
+    virtual void startup() {};  ///< Things to do before handling commands
+    virtual void finishup() {}; ///< things to do when the thread is closing down.
+
+    CommandQueue::Ptr _q{std::make_shared<CommandQueue>()}; ///< Queue of commands.
```

Comment:
_q can be set in the constructor. For ThreadPool, all the threads share the same CommandQueue object. All the schedulers are now based on CommandQueue and passed to the worker ThreadPool when it is made in the SsiService constructor.